hercules-protocol/doc/mapper.md

```diff
@@ -0,0 +1,319 @@
+On deserizlization mapper will expect tag with `Integer` value.
+In case of type mismatch mapper will throw `TypeMismatchException`.
+
+You can redefine default behavior by setting types value:
+
+```java
+class Pojo {
+
+    @Tag(name = "shortTag" types={"Short"})
+    private int value;
+}
+```
+
+In this example on serialization mapper will convert `int` java value to `Short` hercules value
+or throw `InvalidValueException` in case of overflow.
+On deserialization mapper will expect tag of `Short` type and store its value in `int` field.
+
+## Multiple possible types
+
+In Hercules protocol you can set tag to value of different types.
+
+```yaml
+MultitypePojo:
+  multitypeTag: Short, Integer, Long
+```
+
+```java
+class MultitypePojo {
+
+    @Tag(name="multitypeTag", types={"Short", "Integer", "Long"})
+    private long multitypeTag;
+}
+```
+
+All types in `types` annotation parameter must be convertable to java type of field or variable.
+There is standart set of converters from Hercules types to java types:
+
+Byte:
+
+- `Byte` -> `Byte`
+- `Byte` -> `byte`
+- `Byte` -> `Short`
+- `Byte` -> `short`
+- `Byte` -> `Integer`
+- `Byte` -> `int`
+- `Byte` -> `Long`
+- `Byte` -> `long`
+
+Short:
+
+- `Short` -> `Short`
+- `Short` -> `short`
+- `Short` -> `Integer`
+- `Short` -> `int`
+- `Short` -> `Long`
+- `Short` -> `long`
+
+Integer:
+
+- `Integer` -> `Integer`
+- `Integer` -> `int`
+- `Integer` -> `Long`
+- `Integer` -> `long`
+
+Long:
+
+- `Long` -> `Long`
+- `Long` -> `long`
+
+Float:
+
+- `Float` -> `Float`
+- `Float` -> `float`
+- `Float` -> `Double`
+- `Float` -> `double`
+
+Double:
+
+- `Double` -> `Double`
+- `Double` -> `double`
+
+Flag:
+
+- `Flag` -> `Boolean`
+- `Flag` -> `boolean`
+
+String:
+
+- `String` -> `String`
+
+UUID:
+
+- `UUID` -> `UUID`
+
+Null:
+
+- `Null` -> `Byte`
+- `Null` -> `Short`
+- `Null` -> `Integer`
+- `Null` -> `Long`
+- `Null` -> `Float`
+- `Null` -> `Double`
+- `Null` -> `Boolean`
+- `Null` -> `String`
+- `Null` -> `UUID`
+- `Null` -> `Object`
+
+## Immutable class
+
+For immutable classes you must mark one constructor with `@HerculesMapperConstructor` annotation.
+
+The same schema
+
+```yaml
+PojoClass:
+  integerTag: Integer
+  optionalStringTag?: String
+```
+
+will be described as:
+
+```java
+class ImmutablePojoClass {
+
+    private final int integerTag;
+    private final Maybe<String> optionalStringTag;
+
+    @HerculesMapperConstructor
+    public ImmutablePojoClass(
+        @Tag("integerTag") int integerTag,
+        @Tag(name = "optionalStringTag", optional = true) Maybe<String> optionalStringTag
+    ) {
+        this.integerTag = integerTag;
+        this.optionalStringTag = optionalStringTag;
+    }
+
+    // Getters
+}
+```
+
+## Optional tags
+
+In case of optional tag
+
+```yaml
+PojoWithOptionalTag:
+  optionalString?: String
+```
+
+java type must be enclosed in Maybe generic type
+
+```java
+class PojoWithOptionalString {
+
+    @Tag(name = "optionalString", optional = true)
```

Comment:
Also, `@Tag(name = "optionalTag", required = false, defaultValue="N/A")` is similar to `@RequestParam` from Spring MVC ([details](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestParam.html)).

Reply:
It is similar, but defaultValue cannot be null or any object value except string constant. Default value of request parameter can be always described as string, for default tag value this is not true.

Reply:
I suppose that it is not necessary to provide complex default values. Primitives, strings and uuids should be enough.